### PR TITLE
Feature/refactor tools and models

### PR DESCRIPTION
**Pull request by hugohonda**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -29,13 +29,15 @@ controlnet-aux = {version = "^0.0.9", optional = true}
 lmdeploy = {version = "^0.5.3", optional = true}
 decord = {version = "^0.6.0", optional = true}
 sam-2 = {git = "https://github.com/landing-ai/segment-anything-2.git", branch = "main", optional = true}
+flash-attn = "1.0.5"
```
Comment (CamiloInx) [inline pyproject.toml:32]: Installing this dependency is tricky (I've seen in the past that poetry is not capable of handle it correctly). We should mark it as optional.

File: vision_agent_tools/tools/text_to_object_detection.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+class TextToObjectDetectionOutput(BaseModel):
+    output: Any
+
+
+class TextToObjectDetection(BaseTool):
+    """
+    Tool to perform object detection based on text prompts using a specified ML model
+    """
+
+    def __init__(self, model: str):
```
Comment (camiloaz) [inline vision_agent_tools/tools/text_to_object_detection.py:17]: I think it's ok to merge like this, but we should not have model be a generic string. It should be an string `Enum` that defines specific options for each task. For example, `InternLMXComposer2` cannot run this task, but `FlorenceV2` can

File: vision_agent_tools/models/model_registry.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+from typing import Dict, Callable
+from vision_agent_tools.shared_types import BaseMLModel
+from vision_agent_tools.models.florencev2 import Florencev2
+
+
+MODEL_REGISTRY: Dict[str, Callable[[], BaseMLModel]] = {"florence2": Florencev2}
```
Comment (CamiloInx) [inline vision_agent_tools/models/model_registry.py:6]: @hugohonda I think we should include all the models in the `MODEL_REGISTRY`

**Review by camiloaz — APPROVED** [review]
I see that CI is failing because of flash-attn in the pyproject.toml. but apart from that, LGTM.

File: vision_agent_tools/shared_types.py [hunk 4]
```diff
@@ -5,10 +5,28 @@
...
 
 
+class BaseTool:
+    """
+    Base class for all tools that wrap ML models to accomplish tool tasks.
+    Tools are responsible for interfacing with one or more ML models to perform specific tasks.
+    """
+
+    def __init__(self, model: BaseMLModel):
+        self.model = model
```
[the remark this answers is not in the document]
Comment (camiloaz) [reply vision_agent_tools/shared_types.py:24]: Agree, I already discussed this with @hugohonda . @hugohonda let's change the `model` param here from BaseMLModel to str.

File: vision_agent_tools/models/model_registry.py (hunk 3 above)
Comment (camiloaz) [reply vision_agent_tools/models/model_registry.py:6]: I agree. But let's just keep it like this to merge this PR and we can add them in an upcoming PR.

File: pyproject.toml (hunk 1 above)
Comment (hugohonda) [reply pyproject.toml]: Agree! Will add to README on the installation section what is needed to manually install flash-attn:
`poetry run pip install flash-attn`

File: vision_agent_tools/models/model_registry.py (hunk 3 above)
Comment (hugohonda) [reply vision_agent_tools/models/model_registry.py:6]: Ok! Will be added on the upcoming code

**Pull request merged**